x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -6,11 +6,739 @@
+                final int bytesRead = Streams.readFully(encryptedDataInputStream, DEKIdBytes);
+                if (bytesRead != DEK_ID_LENGTH) {
+                    throw new RepositoryException(repositoryName, "The encrypted blob [" + blobName + "] is too small [" + bytesRead + "]");
+                }
+                final String DEKId = new String(DEKIdBytes, StandardCharsets.UTF_8);
+                // might open a connection to read and decrypt the DEK, but most likely it will be served from cache
+                final SecretKey DEK = getDEKById.apply(DEKId);
+                // read and decrypt the rest of the blob
+                return new DecryptionPacketsInputStream(encryptedDataInputStream, DEK, PACKET_LENGTH_IN_BYTES);
+            } catch (Exception e) {
+                try {
+                    encryptedDataInputStream.close();
+                } catch (IOException closeEx) {
+                    e.addSuppressed(closeEx);
+                }
+                throw e;
+            }
+        }
+
+        /**
+         * Reads the blob content from the input stream and writes it to the container in a new blob with the given name.
+         * If {@code failIfAlreadyExists} is {@code true} and a blob with the same name already exists, the write operation will fail;
+         * otherwise, if {@code failIfAlreadyExists} is {@code false} the blob is overwritten.
+         * The contents are encrypted in a streaming fashion. The DEK (encryption key) is randomly generated and reused for encrypting
+         * subsequent blobs such that the same IV is not reused together with the same key.
+         * The DEK encryption key is separately stored in a different blob, which is encrypted with the repository key.
+         *
+         * @param   blobName
+         *          The name of the blob to write the contents of the input stream to.
+         * @param   inputStream
+         *          The input stream from which to retrieve the bytes to write to the blob.
+         * @param   blobSize
+         *          The size of the blob to be written, in bytes. The actual number of bytes written to the storage service is larger
+         *          because of encryption and authentication overhead. It is implementation dependent whether this value is used
+         *          in writing the blob to the repository.
+         * @param   failIfAlreadyExists
+         *          whether to throw a FileAlreadyExistsException if the given blob already exists
+         */
+        @Override
+        public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+            // reuse, but possibly generate and store a new DEK
+            final SingleUseKey singleUseNonceAndDEK = singleUseDEKSupplier.get();
+            final byte[] DEKIdBytes = singleUseNonceAndDEK.getKeyId().getBytes(StandardCharsets.UTF_8);
+            if (DEKIdBytes.length != DEK_ID_LENGTH) {
```

Comment:
> This should be an assertion?

It should probably be an assertion, BUT, as an exception (pun!), I would slightly prefer this case be an `Exception` (a `RepositoryException` wrapping an `IllegalStateException`). I prefer it be an exception because of the `readBlob` and `writeBlob` symmetry; one reads a fixed length prefix while the other writes the prefix. I think it's healthy to keep this redundant check at this level, even if probably the _current code_ would fail in other places in the obverse. An assertion, in my view, signifies that the logic is broken if the assertion is not satisfied, but in this case the logic in the method works OK, but the repository gets corrupted.

> Also maybe just make getKeyId return a BytesReference so we don't have to do the string to byte[] conversion here all the time. It's kind of confusing to handle something that is a bunch of bytes as a string only to convert it to byte[] before every use?

I have pushed d49bf401c44184ec9e996ef468e86d5ec that makes key ids `BytesReference`s instead of `String`s. At first, I thought it was a good suggestion, BUT the Ids were used as strings in `read/writeBlob` methods, in constructing `BlobPath` and in logging. Looking at the code after the change it doesn't look like an improvement to me. Also because now logging does require some processing we have to introduce the message supplier pattern again... I tend to favor the original approach, but, your call.

Reply:
Ah sorry I think there was a misunderstanding here, I was only referring to `getKeyId`, that seems to always just be a `BytesReference` and never be used as `String`? The others should stay `String` if we use them as `String`.

Also makes sense + I'm fine with keeping the exception.

Reply:
Okay, understood, I've pushed 87e64b3dd3bc . Thanks.